flink-core/src/test/java/org/apache/flink/core/fs/LimitedConnectionsFileSystemTest.java

```diff
@@ -122,6 +122,42 @@ public void testLimitingInputStreams() throws Exception {
 		}
 	}
 
+	@Test
+	public void testLimitingRateLimitingStream() throws Exception {
+		final LimitedConnectionsFileSystem limitedFs = new LimitedConnectionsFileSystem(
+				LocalFileSystem.getSharedInstance(),
+				Integer.MAX_VALUE,
+				Integer.MAX_VALUE,
+				Integer.MAX_VALUE,
+				0,
+				0,
+				10000, // Limit write to 10 kbytes/s
+				10000); // Limit read to 10 kbytes/s
+		File file = tempFolder.newFile();
+		Path path = new Path(file.toURI());
+		long durationWrite = System.currentTimeMillis();
+		try (FSDataOutputStream stream = limitedFs.create(path, WriteMode.OVERWRITE)) {
+			final Random rnd = new Random();
+			final byte[] data = new byte[100];
+			for (int i = 0; i < (1000 + 10); i++) {
+				rnd.nextBytes(data);
+				stream.write(data);
+			}
+		}
+		durationWrite = System.currentTimeMillis() - durationWrite;
+
+		long durationRead = System.currentTimeMillis();
+		final byte[] data = new byte[100];
+		try (FSDataInputStream stream = limitedFs.open(path)) {
+			//noinspection StatementWithEmptyBody
+			while (stream.read(data) != -1) {}
+		}
+		durationRead = System.currentTimeMillis() - durationRead;
+		file.delete();
+		assertTrue(durationWrite > 10000);
+		assertTrue(durationRead > 8000); // Less stability with read limiter than write
```

Comment:
I have debugged it, and my guess is that this instability came from bursty reads (`org.apache.flink.shaded.guava18.com.google.common.util.concurrent.SmoothRateLimiter.SmoothBursty`). Default `RateLimiter` comes with 1 second of max burst duration and this "burst credits" are accumulated for reads while this test is writing. I have modified this test to take that into account (and added test case for single byte writes/reads), please check the code below:

```

	@Test
	public void testLimitingRateLimitingStream() throws Exception {
		testLimitingRateLimitingStream(10);
	}

	@Test
	public void testLimitingRateLimitingStreamSingeByteWrites() throws Exception {
		testLimitingRateLimitingStream(1);
	}

	private void testLimitingRateLimitingStream(int chunkSize) throws Exception {
		long dataRate = 1000; // Limit read to 1000 bytes/s
		// Allowed burst duration in SmoothBursty when creating RateLimiter. This value is copied from `RateLimiter#create(SleepingStopwatch, double)`
		long burstDurationSeconds = 1; 
		long burstData = dataRate * burstDurationSeconds;
		long dataToWrite = 2000;
		assertTrue(
			"Becauese of burst writes/reads, dataToWrite must be larger then burstData",
			dataToWrite > burstData);
		final LimitedConnectionsFileSystem limitedFs = new LimitedConnectionsFileSystem(
			LocalFileSystem.getSharedInstance(),
			Integer.MAX_VALUE,
			Integer.MAX_VALUE,
			Integer.MAX_VALUE,
			0,
			0,
			dataRate,
			dataRate);
		File file = tempFolder.newFile();
		Path path = new Path(file.toURI());
		long durationWrite = System.currentTimeMillis();
		try (FSDataOutputStream stream = limitedFs.create(path, WriteMode.OVERWRITE)) {
			final byte[] data = new byte[chunkSize];
			for (int i = 0; i < dataToWrite; i += chunkSize) {
				if (chunkSize == 1) {
					stream.write(42);
				}
				else {
					stream.write(data);
				}
			}
		}
		durationWrite = System.currentTimeMillis() - durationWrite;

		long durationRead = System.currentTimeMillis();
		try (FSDataInputStream stream = limitedFs.open(path)) {
			final byte[] data = new byte[chunkSize];
			//noinspection StatementWithEmptyBody
			int result = 0;
			while (result != -1) {
				result = chunkSize == 1 ? stream.read() : stream.read(data);
			}
		}
		durationRead = System.currentTimeMillis() - durationRead;
		file.delete();
		long expectedDuration = (long) (((dataToWrite - burstData) * 1000) / dataRate * (0.9));
		assertThat(durationWrite, greaterThan(expectedDuration));
		assertThat(durationRead, greaterThan(expectedDuration));
	}
```

Reply:
I agree with the burst behaviour of guava RateLimiter but I don't explain why we don't have symmetric behaviour between read and write : 
* For the write I am always above the 10s (in my test)
* For the read I am sometimes below 
I played differents tests yesterday and don't have clues about the differences.

Reply:
I think burst credits are accumulated over time. If you create RateLimiter in `LimitedConnectionsFileSystem` constructor and immediately start using the output rate limiter, it will not have any burst credits accumulated. However input rate limiter will be used only after couple of seconds (when write has finished), thus it will accumulate burst credits.

Btw, that makes me thing, do we really want to have two separate limiters for read and writes? Are the quotas on S3 accounted also separately? Or usually is there a single quota for reads and writes together? If the latter one, it would be more handy to have one shared `RateLimiter` between reads and writes.

Reply:
My initial need is only for the output Ratelimiter :  we want to be able to limit the rate of sending checkpoint to s3/hadoop because it is in concurrence with flink exchange of data between task manager (we are on an infrastructure with only gigabit) 
I implemented input and output because some people could also want to limit input . (even if I don't see real use cases for input ratelimiter)
I implemented separated limiter because on a full duplex network it is possible you want to privilege one part of the traffic but I am also ok to have a shared ratelimiter

Reply:
Ok, I'm fine to keep it as it is.

Reply:
@etiennecarriere could you apply the test changes that I suggested above, so that we could merge this PR?